File: leetCode/python/p34_FindFirstAndLastPositionOfElementInSortedArray/lc34.py
#!/usr/bin/python3

class Solution:
    '''
        Given an array of integers nums sorted in non-decreasing order, 
        find the starting and ending position of a given target value.

        If target is not found in the array, return [-1, -1].

        You must write an algorithm with O(log n) runtime complexity.
    '''
    
    def searchRange(self, nums: list, target: int) -> list:
        
        start = 0
        end = len(nums) - 1
        sol = [-1, -1]

        if len(nums) == 0:
            return sol

        while start <= end:
            
            mid = (start + end) // 2;
        
            if nums[mid] == target:
                
                sol[0] = mid - 1
                temp = mid

                while nums[mid] == nums[temp]:
                    temp += 1

                sol[1] = temp - 1

                break
            elif nums[mid] < target:
                start = mid + 1
            else:
                end = mid - 1

        return sol

if __name__ == '__main__':

    nums = [5, 7, 7, 8, 8, 10]

    s = Solution()

    # Test 1:

    target = 8

    print('\nTest 1:\n') 
    print('List 1:', end = ' ')
    print(nums)
    print('Target 1: {}'.format(target))
    print('Solution 1:', end = ' ')
    print(s.searchRange(nums, target))

    # Test 2:
     
    target = 6

    print('\nTest 2:\n') 
    print('List 2:', end = ' ')
    print(nums)
    print('Target 2: {}'.format(target))
    print('Solution 2:', end = ' ')
    print(s.searchRange(nums, target))

    # Test 3:

    nums = []
    target = 0

    print('\nTest 3:\n') 
    print('List 3:', end = ' ')
    print(nums)
    print('Target 3: {}'.format(target))
    print('Solution 3:', end = ' ')
    print(s.searchRange(nums, target))

File: leetCode/c/p34_FindFirstAndLastPositionOfElementInSortedArray/lc34.c
/* Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.

If target is not found in the array, return [-1, -1]. */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "../../../mylib/mylib.h"

void searchRange(int* nums, int numsSize, int target, int* sol) {
    
    int start = 0;
    int end = numsSize - 1;
    int mid;
    sol[0] = -1;
    sol[1] = -1;

    if(numsSize == 0) {
        return;
    }

    while(start <= end) {
        
        mid = (start + end) / 2;

        if(nums[mid] == target) {
            
            sol[0] = mid - 1;
            int temp = mid;
        
            while(nums[mid] == nums[temp])
                temp++;

            sol[1] = temp - 1;

            break;
        }
        else if(nums[mid] < target)
            start = mid + 1;
        else
            end = mid - 1; 
    }        
}

int main(void) {

    int nums[6] = {5, 7, 7, 8, 8, 10}; 
    int numsSize = sizeof(nums) / sizeof(int);
    int sol[2] = {-1, -1};
    
    // Test 1:

    printf("\nTest 1:\n"); 

    int target1 = 8;

    printf("Array 1: "); 
    printArrayInt(nums, 6);
    printf("Target 1: %d\n", target1);
    searchRange(nums, numsSize, target1, sol);
    printf("Solution 1: ");
    printArrayInt(sol, 2); 

    // Test 2:

    printf("\nTest 2:\n"); 

    int target2 = 6;

    printf("Array 2: ");
    printArrayInt(nums, 6);
    printf("Target 2: %d\n", target2);
    searchRange(nums, numsSize, target2, sol);
    printf("Solution 2: ");
    printArrayInt(sol, 2); 

    // Test 3:

    printf("\nTest 3:\n"); 
   
    int target3 = 0;
    int* nums1;
    allocateArrayInt(nums1, 0);
 
    printf("Array 3: Empty\n");
    printf("Target 3: %d\n", target3);
    searchRange(nums1, 0, target3, sol);
    printf("Solution 3: ");
    printArrayInt(sol, 2); 

    return EXIT_SUCCESS;
}
